src/mailer/engine.py

```diff
@@ -118,6 +118,18 @@ def _throttle_emails():
         time.sleep(EMAIL_THROTTLE)
 
 
+def handle_backend_exception(connection, message, err):
+    # default error handler
+    message.defer()
+    logging.info("message deferred due to failure: %s" % err)
+    MessageLog.objects.log(message, RESULT_FAILURE, log_message=str(err))
+    action = 'deferred'
+    # Kill the connection, in case the connection itself has an error.
+    connection = None
+
+    return connection, action
```

Comment:
I think the handler here should re-raise the exception if it is not an instance of `(socket_error, smtplib.SMTPSenderRefused, smtplib.SMTPRecipientsRefused, smtplib.SMTPDataError, smtplib.SMTPAuthenticationError)`. The reasons for this are:

1. Preserve the existing behaviour - previously we were only catching these exceptions, not all `Exception`
2. Broad exceptions mask developer errors, and make debugging harder.
3. The best thing to do with unexpected errors here is to crash. This will result in a loud failure (a failed task), rather than a quite one (deferred mail). 

The only disadvantage here is that if you just crash out on unexpected errors, later messages in the queue are not even attempted. However, this will probably be better than a developer error triggering this broad exception and causing all messages to just be put as 'deferred'.

With that in mind, we might want to:
1. change the name of the default handler to make it clear that it does SMTP specific handling,
2. add additional notes to the docs saying that if you change the default backend you will likely need to change the error handling
3. use the default error handler as an example.

Finally, I think `exc` or `exception` would be a better name for the 3rd parameter, and a docstring here would be good.

Reply:
> I think the handler here should re-raise the exception if it is not an instance of `(socket_error, smtplib.SMTPSenderRefused, smtplib.SMTPRecipientsRefused, smtplib.SMTPDataError, smtplib.SMTPAuthenticationError)`. The reasons for this are:
> 
>     1. Preserve the existing behaviour - previously we were only catching these exceptions, not all `Exception`
> 
>     2. Broad exceptions mask developer errors, and make debugging harder.
> 
>     3. The best thing to do with unexpected errors here is to crash. This will result in a loud failure (a failed task), rather than a quite one (deferred mail).
> 
> 
> The only disadvantage here is that if you just crash out on unexpected errors, later messages in the queue are not even attempted. However, this will probably be better than a developer error triggering this broad exception and causing all messages to just be put as 'deferred'.

I get the general idea, I think you have a point here :+1: 

> With that in mind, we might want to:
> 
>     1. change the name of the default handler to make it clear that it does SMTP specific handling,

I think backend connections could be anything not just SMTP: am I missing something?


>     2. add additional notes to the docs saying that if you change the default backend you will likely need to change the error handling

:+1: 

>     3. use the default error handler as an example.

If the default handler just re-raises, I don't think it will make a good example for a handler. It maybe help understand the default behavior but then the doc-copy of that code will have to be in tight sync. Not sure if the pluses outweight the minuses here. What do you think?

> Finally, I think `exc` or `exception` would be a better name for the 3rd parameter, and a docstring here would be good.

:+1:

Reply:
@spookylukey I thought about this more. There are two things I would like to point out:

I just re-read the description of issue #73 and this sentence caught my eye:

>  I realize broad catch statements are generally a bad idea, but this seems like a case where it might be appropriate.

Are you aware that when we re-raise exceptions that user code may need to "see" an instance of such an exception in practice *first* — potentially in production —, to know they have a practical problem when they don't implement a custom error handler.
I have a feeling we might have a damned-if-you-do-damned-if-you-dont situation here or that at least we'd need to resolve a conflict in requirements here. No?

For the second thing, I do appreciate your feedback and the discussion here. At the same time, you explicitly offered at https://github.com/pinax/django-mailer/issues/73#issuecomment-708980819 to @volksman and me:

> So, once you two have consensus that the work is done, I'll merge the PR.

I'm happy to keep discussing it but at the same time I think it will be hard now to expect @volksman to do a shift in direction here, with that agreement made before. What I can offer, if it helps and is welcome to both of you, is to build upon @volksman's current commits and take over the implementer's part from here in a new pull request while you and @volksman take a seat at the review table. All I would need for that is that you two then find eventual consensus and conflict-free requirements and that review continues to happen in a timely manner. I hope that's an offer of value but it's just an idea so if there are other ways forward I'm happy to see how I can contribute.

@volksman @spookylukey please let me know what you think :pray:

Reply:
> If the default handler just re-raises, I don't think it will make a good example for a handler. It maybe help understand the default behavior but then the doc-copy of that code will have to be in tight sync. Not sure if the pluses outweight the minuses here. What do you think?

I was thinking that the default handler should look like this:

```python
def handle_backend_exception(connection, message, exc):
    if not isinstance(exc, (socket_error, smtplib.SMTPSenderRefused, smtplib.SMTPRecipientsRefused, smtplib.SMTPDataError, smtplib.SMTPAuthenticationError)):
        six.reraise(*sys.exc_info())
    message.defer()
    logging.info("message deferred due to failure: %s" % exc)
    MessageLog.objects.log(message, RESULT_FAILURE, log_message=str(exc))
    action = 'deferred'
    # Kill the connection, in case the connection itself has an error.
    connection = None

    return connection, action
```

The reason is maintaining existing behaviour. I agree that this is a "damned-if-you-do-damned-if-you-dont" situation here, and that's why "defaulting to keeping existing behaviour" is a helpful tie-breaker.

With that change in mind, the function would be better named `handle_backend_exception_smtp` or something, since it is SMTP specific.

I think we could also include the current  `handle_backend_exception` as it is (i.e. catching and not reraising all `Exceptions`) as a documented alternative. If my proposed `handle_backend_exception_smtp` is not the default, I think it is important that it is included, with a note in the changelog, so that there is a very easy upgrade path for people who just want to maintain existing functionality.

@hartwork I'm happy to accept your offer of building on the PR and addressing this final issue. I think the finish line is in sight!